README.md
# Power League [![GoDoc](https://godoc.org/github.com/Forestmb/power-league?status.png)](https://godoc.org/github.com/Forestmb/power-league) #

Power League is a web application that calculates alternative rankings for
Yahoo Fantasy Sports leagues.

This application is written using the Go programming language and is licensed
under the [New BSD license](
https://github.com/Forestmb/power-league/blob/master/LICENSE).

## Running ##

The `power-league` application is distributed as a [docker container](
https://github.com/users/Forestmb/packages/container/package/power-league) and
can be run with any compatible container runtime.

To run the server, first obtain a [Yahoo Fantasy Sports client key and secret](
http://developer.yahoo.com/fantasysports/guide/GettingStarted.html). When
registering you will be asked to enter in a redirect URL, which should be in the
format `https://[hostname]/[base-context]/auth`. The `[hostname]` is the
hostname or IP address of the machine where you are running the server, and
`[base-context]` is the optional additional context the application is published
to (defaults to none).

For running on your local machine you will likely need to use a redirect URL of
`https://127.0.0.1/auth`. As of September 2020 you cannot register an
application with any of the following in the redirect URL:

- An `http://` protocol
- A `localhost` hostname
- A non-default port (not 443)

Once your application is registered you can pass the client key/secret into the
application by setting the environment variables `OAUTH_CLIENT_KEY` and
`OAUTH_CLIENT_SECRET` and passing them when running docker:

    $ docker run --rm \
             -p443:443 \
             -e OAUTH_CLIENT_KEY \
             -e OAUTH_CLIENT_SECRET \
             ghcr.io/forestmb/power-league:latest

After it has started, browse to `https://127.0.0.1/` to access the application..
By default a self-signed test certificate will be used. To provide your own TLS
certificates you can mount them using docker volumes and with additional
command-line flags:

    $ docker run --rm \
             -p443:443 \
             -e OAUTH_CLIENT_KEY \
             -e OAUTH_CLIENT_SECRET \
             -v /path/to/cert.pem:/app/certs/cert.pem \
             -v /path/to/key.pem:/app/certs/key.pem \
             ghcr.io/forestmb/power-league:latest \
             -tlsCert /app/certs/cert.pem \
             -tlsKey /app/certs/key.pem

If you are running the application behind a reverse proxy like nginx or Apache
and do not need or wish to run with HTTPS, you can opt out when running the
server. However when registering your application you will still need to provide
an HTTPS redirect URL, and you should pass in the publicly-accessible redirect
URL when starting the container:

    $ docker run --rm \
             -p8080:8080 \
             -e OAUTH_CLIENT_KEY \
             -e OAUTH_CLIENT_SECRET \
             ghcr.io/forestmb/power-league:latest \
             -noTLS \
             -address :8008 \
             -redirectURL https://public.url.example.com/auth

Once your server is running, visit it in the web browser and sign in. After
granting access you can view the rankings for any of your current or past
leagues:

![Example Screenshot](https://raw.github.com/Forestmb/power-league/master/doc/screenshots/rankings.png)

## Building ##

Building requires an either a [Docker](https://www.docker.com/) compatible
container build engine or an installation of the [Go programming language tools](
https://golang.org/doc/install). To build:

    # Build using docker
    $ docker build -t power-league:latest .

    # Build server locally
    $ ./build.sh

To run the local build before every commit, use:

    $ ln -s "$(pwd)/build.sh" .git/hooks/pre-commit

## Options ##

Command line flags can be passed when running either locally or as additional docker run
command arguments.

    Usage of ./power-league:
      -address string
        	Address to listen for incoming connections. (default ":443")
      -alsologtostderr
        	log to standard error as well as files
      -baseContext string
        	Root context of the server. (default "/")
      -clientKey string
        	Required client OAuth key. Defaults to the value of OAUTH_CLIENT_KEY.
            See http://developer.yahoo.com/fantasysports/guide/GettingStarted.html
            for more information
      -clientSecret string
        	Required client OAuth secret. Defaults to the value of
            OAUTH_CLIENT_SECRET.
            See http://developer.yahoo.com/fantasysports/guide/GettingStarted.html
            for more information
      -cookieAuthKey string
        	Authentication key for cookie store. Defaults to the value of
            COOKIE_AUTH_KEY. By default uses a randomly generated key.
      -cookieEncryptionKey string
        	Encryption key for cookie store. Defaults to the value of
            COOKIE_ENCRYPTION_KEY. By default uses a randomly generated key.
      -log_backtrace_at value
        	when logging hits line file:N, emit a stack trace
      -log_dir string
        	If non-empty, write log files in this directory
      -logtostderr
        	log to standard error instead of files
      -minimizeAPICalls
        	Minimize calls to the Yahoo Fantasy Sports API. If enabled, it will
            lower the risk of being throttled but will result in a higher
            average page load time.
      -noTLS
        	Disable TLS.
      -static string
        	Directory to access static files (default "static")
      -stderrthreshold value
        	logs at or above this threshold go to stderr
      -tlsCert string
        	TLS certificate if using HTTPS. (default "./certs/localhost.crt")
      -tlsKey string
        	TLS private key if using HTTPS. (default "./certs/localhost.key")
      -totalCacheSize int
        	Maximum number of responses that well be cached across all users.
            (default 10000)
      -trackingID string
        	Google Analytics tracking ID. If blank, tracking will not be
            activated. Defaults to value of the GA_TRACKING_ID environment
            variable.
      -userCacheDurationSeconds int
        	Maximum duration user data will be cached, in seconds. Defaults to
            six hours (default 21600)
      -v value
        	log level for V logs
      -vmodule value
        	comma-separated list of pattern=N settings for file-filtered logging


CHANGELOG.md
# Changelog #

## 0.4.0 (2020-09-20) ##

- Lower required number of API calls (disabled by default).
- Made cache size configurable and lowered default to 10000.
- Added graphs to show weekly power rank and fantasy points over the course
  of the season.
- Improved mobile authorization flow.
- Added 'Victory Points' and 'Total Points' ranking schemes.
- OAuth 2.0 for authentication.
- Switched to containers as primary means of running/deploying.
- Added go mod support.

## 0.3.0 (2015-01-09) ##

- Traffic analysis through Google Analytics.
- Added user preference for displaying all-play records instead of power scores.
- Added column sorting to rankings page.
- Added league rank and record to rankings page.
- Overall rankings can now be exported to CSV or a newsletter template.

## 0.2.0 (2014-09-05) ##

- Power ranking projections for unfinished leagues.
- Caches fantasy content for each user for up to 6 hours (configurable).
- Fixed error message when displaying leagues pre-draft.

## 0.1.0 (2014-03-02) ##

Initial public release

- OAuth 1.0 for authentication.
- Support for fantasy football leagues.


power-league.go
// Command power-league starts a web server that display alternative rankings
// for fantasy sports leagues.
//
// Users grant authentication from the fantasy sports provider (Yahoo) and
// are presented with a list of their current and past leagues. Each league
// contains a table and week-by-week breakdown of how their scores translate
// into the alternative rankings.
//
// To run, this command requires a client key and secret to be passed in at
// runtime. More information about the registration process can be found here:
// http://developer.yahoo.com/fantasysports/guide/GettingStarted.html
//
// Additional configuration options are available. See usage for details.
package main

import (
	"flag"
	"fmt"
	"net/http"
	"os"

	"github.com/Forestmb/goff"
	"github.com/Forestmb/power-league/rankings"
	"github.com/Forestmb/power-league/session"
	"github.com/Forestmb/power-league/site"
	"github.com/golang/glog"
	"github.com/gorilla/handlers"
	"github.com/gorilla/securecookie"
	"github.com/gorilla/sessions"
)

func main() {
	addr := flag.String(
		"address",
		":443",
		"Address to listen for incoming connections.")
	noTLS := flag.Bool("noTLS", false, "Disable TLS.")
	tlsCert := flag.String("tlsCert", "./certs/localhost.crt", "TLS certificate if using HTTPS.")
	tlsKey := flag.String("tlsKey", "./certs/localhost.key", "TLS private key if using HTTPS.")
	baseContextFlag := flag.String(
		"baseContext",
		"/",
		"Root context of the server.")
	staticFilesLocation := flag.String(
		"static",
		"static",
		"Directory to access static files")
	userCacheDurationSeconds := flag.Int(
		"userCacheDurationSeconds",
		6*60*60,
		"Maximum duration user data will be cached, in seconds. Defaults to"+
			" six hours")
	totalCacheSize := flag.Int64(
		"totalCacheSize",
		10000,
		"Maximum number of responses that well be cached across all users.")
	minimizeAPICalls := flag.Bool(
		"minimizeAPICalls",
		false,
		"Minimize calls to the Yahoo Fantasy Sports API. If enabled, it will "+
			"lower the risk of being throttled but will result in a higher "+
			"average page load time.")
	trackingID := flag.String(
		"trackingID",
		os.Getenv("GA_TRACKING_ID"),
		"Google Analytics tracking ID. If blank, tracking will not be activated. "+
			"Defaults to value of the GA_TRACKING_ID environment variable.")
	clientKey := flag.String(
		"clientKey",
		"",
		"Required client OAuth key. Defaults to the value of OAUTH_CLIENT_KEY. "+
			"See http://developer.yahoo.com/fantasysports/guide/GettingStarted.html"+
			" for more information")
	clientSecret := flag.String(
		"clientSecret",
		"",
		"Required client OAuth secret. Defaults to the value of OAUTH_CLIENT_SECRET. "+
			"See http://developer.yahoo.com/fantasysports/guide/GettingStarted.html"+
			" for more information")
	clientRedirectURL := flag.String(
		"clientRedirectURL",
		"",
		"Client redirect URL. Defaults to the listening address.")
	cookieAuthKey := flag.String(
		"cookieAuthKey",
		"",
		"Authentication key for cookie store. Defaults to the value of COOKIE_AUTH_KEY. "+
			"By default uses a randomly generated key.")
	cookieEncryptionKey := flag.String(
		"cookieEncryptionKey",
		"",
		"Encryption key for cookie store. Defaults to the value of COOKIE_ENCRYPTION_KEY. "+
			"By default uses a randomly generated key.")
	flag.Parse()
	defer glog.Flush()

	invalidInputParameters := false
	if *clientKey == "" {
		envValue := os.Getenv("OAUTH_CLIENT_KEY")
		clientKey = &envValue
	}
	if *clientKey == "" {
		fmt.Fprintln(os.Stderr, "power-league: clientKey must be provided")
		invalidInputParameters = true
	}

	if *clientSecret == "" {
		envValue := os.Getenv("OAUTH_CLIENT_SECRET")
		clientSecret = &envValue
	}
	if *clientSecret == "" {
		fmt.Fprintln(os.Stderr, "power-league: clientSecret must be provided")
		invalidInputParameters = true
	}

	if !*noTLS {
		if *tlsCert == "" {
			fmt.Fprintln(os.Stderr, "power-league: tlsCert must be provided")
			invalidInputParameters = true
		}
		if *tlsKey == "" {
			fmt.Fprintln(os.Stderr, "power-league: tlsKey must be provided")
			invalidInputParameters = true
		}
	}

	if invalidInputParameters {
		os.Exit(1)
	}

	// Remove trailing slashes
	baseContext := *baseContextFlag
	size := len(baseContext)
	for size > 0 && baseContext[size-1] == '/' {
		baseContext = baseContext[:size-1]
		size = size - 1
	}
	glog.Infof("starting power rankings site -- context=%s", baseContext)

	rankings.MinimizeAPICalls = *minimizeAPICalls

	// Create cookie store
	var cookieStoreAuthKey []byte
	if len(*cookieAuthKey) == 0 {
		envValue := os.Getenv("COOKIE_AUTH_KEY")
		cookieAuthKey = &envValue
	}
	if len(*cookieAuthKey) == 0 {
		glog.V(2).Infoln("using randomly generated cookie authentication key")
		cookieStoreAuthKey = securecookie.GenerateRandomKey(32)
	} else {
		glog.V(2).Infoln("using cookie authentication key from command line")
		cookieStoreAuthKey = []byte(*cookieAuthKey)
	}

	var cookieStoreEncryptionKey []byte
	if len(*cookieEncryptionKey) == 0 {
		envValue := os.Getenv("COOKIE_ENCRYPTION_KEY")
		cookieEncryptionKey = &envValue
	}
	if len(*cookieEncryptionKey) == 0 {
		glog.V(2).Infoln("using randomly generated cookie encryption key")
		cookieStoreEncryptionKey = securecookie.GenerateRandomKey(32)
	} else {
		glog.V(2).Infoln("using cookie encryption key from command line")
		cookieStoreEncryptionKey = []byte(*cookieEncryptionKey)
	}

	authContext := fmt.Sprintf("%s/auth", baseContext)
	sessionManager := session.NewManagerWithCache(
		oauth2ConsumerProvider{
			tls:          !*noTLS,
			clientKey:    *clientKey,
			clientSecret: *clientSecret,
			redirectURL:  *clientRedirectURL,
			authContext:  authContext,
		},
		sessions.NewCookieStore(cookieStoreAuthKey, cookieStoreEncryptionKey),
		*userCacheDurationSeconds,
		*totalCacheSize)

	site := site.NewSite(
		!*noTLS, baseContext, *staticFilesLocation, "templates/html/", *trackingID, sessionManager)
	var err error
	if *noTLS {
		err = http.ListenAndServe(*addr, handlers.LoggingHandler(logWriter{}, site.ServeMux))
	} else {
		err = http.ListenAndServeTLS(*addr, *tlsCert, *tlsKey, handlers.LoggingHandler(logWriter{}, site.ServeMux))
	}
	if err != nil {
		glog.Exit("ListenAndServe: ", err)
	}
}

// logWriter implements io.Writer to write to the correct logging
// implementation.
type logWriter struct{}

// Write the input to the correct logging implementation
func (l logWriter) Write(p []byte) (int, error) {
	glog.Infof(string(p))
	return len(p), nil
}

// oauth2ConsumerProvider implements session.Consumer to provide the OAuth 2
// config for this application
type oauth2ConsumerProvider struct {
	tls          bool
	clientKey    string
	clientSecret string
	redirectURL  string
	authContext  string
}

func (o oauth2ConsumerProvider) Get(r *http.Request) session.Consumer {
	redirectURL := o.redirectURL
	if redirectURL == "" {
		protocol := "https"
		if !o.tls {
			protocol = "http"
		}
		redirectURL = fmt.Sprintf("%s://%s%s", protocol, r.Host, o.authContext)
	}
	return goff.GetOAuth2Config(o.clientKey, o.clientSecret, redirectURL)
}
